docs/development/vscode.md

```diff
@@ -0,0 +1,143 @@
+
+    ```python
+    pip install ptvsd
+    ```
+
+8. Edit `tools/settings.py`
+
+    Add the following lines in `zproject/settings.py` after `if DEBUG:`
+    ```python
+    if DEBUG:                                   # Already present
+        INTERNAL_IPS = ('127.0.0.1',)           # Already Present
+        import ptvsd
+        if "runserver" in sys.argv and os.environ.get("RUN_MAIN"):
+            try:
+                ptvsd.enable_attach(redirect_output=True)
+                print("VS Code Debugging Enabled.")
+            except OSError:
+                print("Debugging Setup Error, Exception: ")
+                tracback.print_exc()
+    ```
+
+    The `redirect_output` gets the output from the server redirected to
+    the debug console of VS Code (quite a handy thing!).
+
+    Here you can have a bit of modifications like:
+    - If you want the django server to wait for debugger *prior to
+      running* just add
+    ```python
+    ...
+    ptvsd.enable_attach()
+    ptvsd.wait_for_debugger()
+    ...
+    ```
+   - If you want to run the ptvsd server on some other port than default
+     (default is *0.0.0.0* and *5678* ) then
+    ```python
+    ptvsd.enable_attach(YOUR_HOST_IP, YOUR_HOST_PORT)
+    ```
+
+9. Create a `launch.json` file
+
+    Add a debugging config by checking out the instructions
+    [here](https://code.visualstudio.com/docs/editor/debugging#_launch-configurations).
+    In the launch.json add the following lines
+
+    ```js
+    {
+        "version": "0.2.0",
+        "configurations": [
+            {
+            "name":"Python: Remote Attach",     // name of config
+            "type":"python",
+            "request":"attach",                 // attach type request
+            "port":5678,                        // the port on which ptvsd is running
+            "host":"localhost",                 // host ip of the vagrant server
+            "django" : true,                    // Not neccessary
+            "subprocess" : true,                // multi-process architecture
+            "pathMappings":[{                   // your local->server mappinps
+                "localRoot":"${workspaceFolder}",
+                "remoteRoot":"."
+                }]
+            }  
+        ]
+    }
+    ```
+
+10. Launch `run-dev.py` via command line
+11. Attach the debugger from VS Code
+12. Set up breakpoints and catch those nasty bugs!
+
+## Note
+- You need to attach the debugger everytime the sever reloads. No
+  alternative to this right now.
+- You need to unstage `settings.py`, `Vagrantfile` and `Dockerfile`
+  before committing.
+- Haven't tested with manual installation of Zulip
+- Need an internet connection
+- Not secure for running on production servers and / or for remote dev
+  server (i.e. not on local servers).
+- You may need to increase the nginx's `connection_timeout` and
+  `read_timeout` for stepping through the code for long times and still
+  needing a response at client side.
```

Comment:
I don't understand this one -- there's no `nginx` in the development environment...

Reply:
Well, yeah I think I should remove that. I mistook that from [here](https://zulip.readthedocs.io/en/latest/overview/architecture-overview.html#components). But I guess it's for production? 

The point however was, as we are debugging for more time say > 1-2 mins. As the client is waiting for a response from the server (which is stuck in debugging at the moment), a timeout occurs and no response is shown on the client side. Just a reminder for the same.

Reply:
Yeah, this is only an issue in production, and we can remove it here.